code/parachain/frame/cosmwasm/src/lib.rs

```diff
@@ -498,1084 +513,836 @@ pub mod pallet {
-		let label: ContractLabelOf<T> = label
-			.try_into()
-			.map_err(|_| CosmwasmVMError::Rpc(String::from("'label' is too large")))?;
-		let funds: FundsOf<T> = funds
-			.try_into()
-			.map_err(|_| CosmwasmVMError::Rpc(String::from("'funds' is too large")))?;
-		let message: ContractMessageOf<T> = message
-			.try_into()
-			.map_err(|_| CosmwasmVMError::Rpc(String::from("'message' is too large")))?;
-		let mut shared = Pallet::<T>::do_create_vm_shared(gas, InitialStorageMutability::ReadWrite);
-		setup_instantiate_call(instantiator, code_id, &salt, admin, label, &message)?.call(
-			&mut shared,
-			funds,
-			message,
+	pub(crate) fn do_continue_reply<'a>(
+		vm: &'a mut DefaultCosmwasmVM<T>,
+		reply: cosmwasm_vm::cosmwasm_std::Reply,
+		event_handler: &mut dyn FnMut(cosmwasm_vm::cosmwasm_std::Event),
+	) -> Result<Option<cosmwasm_vm::cosmwasm_std::Binary>, CosmwasmVMError<T>> {
+		setup_reply_call(
+			vm.contract_address.clone().into_inner(),
+			vm.contract_address.clone().into_inner(),
+		)?
+		.sub_call(
+			vm.shared,
+			Vec::default(),
+			&serde_json::to_vec(&reply).map_err(|_| Error::<T>::FailedToSerialize)?,
+			event_handler,
 		)
 	}
 
-	pub fn execute<T: Config>(
-		executor: AccountIdOf<T>,
+	pub(crate) fn do_continue_migrate<'a>(
+		vm: &'a mut DefaultCosmwasmVM<T>,
 		contract: AccountIdOf<T>,
-		funds: FundsOf<T>,
-		gas: u64,
-		message: ContractMessageOf<T>,
-	) -> Result<(), CosmwasmVMError<T>> {
-		let mut shared = Pallet::<T>::do_create_vm_shared(gas, InitialStorageMutability::ReadWrite);
-		setup_execute_call(executor, contract)?.call(&mut shared, funds, message)
+		message: &[u8],
+		event_handler: &mut dyn FnMut(cosmwasm_vm::cosmwasm_std::Event),
+	) -> Result<Option<cosmwasm_vm::cosmwasm_std::Binary>, CosmwasmVMError<T>> {
+		let CosmwasmContractMeta { code_id, .. } = Self::do_running_contract_meta(vm);
+		setup_migrate_call(vm.shared, vm.contract_address.clone().into_inner(), contract, code_id)?
+			.sub_call(vm.shared, Default::default(), message, event_handler)
+	}
+
+	pub(crate) fn do_query_info(
+		vm: &mut DefaultCosmwasmVM<T>,
+		address: AccountIdOf<T>,
+	) -> Result<ContractInfoResponse, CosmwasmVMError<T>> {
+		// TODO: cache or at least check if its current contract and use `self.contract_info`
+		let (contract_info, code_info) = match T::PalletHook::info(&address) {
+			Some(precompiled_info) => (precompiled_info.contract, precompiled_info.code),
+			None => {
+				let contract_info = if &address == vm.contract_address.as_ref() {
+					vm.contract_info.clone()
+				} else {
+					Pallet::<T>::contract_info(&address)?
+				};
+				let code_info = CodeIdToInfo::<T>::get(contract_info.code_id)
+					.ok_or(Error::<T>::CodeNotFound)?;
+				(contract_info, code_info)
+			},
+		};
+		let ibc_port = if code_info.ibc_capable {
+			Some(Pallet::<T>::do_compute_ibc_contract_port(address))
+		} else {
+			None
+		};
+		let pinned = vm.shared.cache.code.contains_key(&contract_info.code_id);
```

Comment:
We can return pinned from the above match and have a constant `true` for pallets.